### ISSUE TITLE
Support UPSERT in LanceDB

### ISSUE THREAD

**Issue by eddyxu** [lancedb/lancedb#524]
# Problem Statement

Support upsert in LanceDB to allow user update data

**Comment by wjones127** [comment]
Can we do something similar to what we did with `update` and have a temporary implementation that is delete + append?

https://github.com/lancedb/lancedb/blob/d326146a403314528d36ad6a2cc8eed90b51f12b/python/lancedb/table.py#L795

**Comment by prrao87** [comment]
Delete + append will work well in the interim, but I think having an upsert is definitely of huge value in the real world!

A common use case where delete + append isn't very useful is when you have an existing table with 1M vectors, and you receive a new (updated) batch of raw data of a much smaller size, say, 50K records. Not all of these 50K records exist in the database, but some of them do, and we don't know which ones do (unless we do a full scale on the DB of 1M records) to generate an intersection ID list. Classic situation one faces everyday in Postgres of MongoDB. In such cases, matching on an ID field and switching between update/add at the Rust level would be hugely helpful, as opposed to users attempting to do this via the Python/JS API.

I realize that Lance doesn't have the concept of primary keys, but maybe it's possible to circumvent that requirement by building some sort of lookup index for a specified field (like ID) for the upsert operation to be more efficient? 🤔

**Comment by wjones127** [comment]
> I think having an upsert is definitely of huge value in the real world!

Definitely, upsert/merge is 100% on the roadmap for Lance and LanceDB. There's just a lot on that roadmap, so having an interim solution would be nice to take the pressure off until we can get to a proper implementation.

**Comment by gsakkis** [comment]
> Delete + append will work well in the interim

I tried to use this in a multithreaded context but delete (unlike append) is not thread-safe:
```py
  File "/opt/mambaforge/envs/vdb/lib/python3.10/site-packages/lancedb/table.py", line 793, in delete
    self._dataset.delete(where)
  File "/opt/mambaforge/envs/vdb/lib/python3.10/site-packages/lance/dataset.py", line 533, in delete
    self._ds.delete(predicate)
RuntimeError: Already borrowed
```

**Comment by wjones127** [comment]
@gsakkis Thanks for reporting that. We'll probably want to change that to be thread safe.

**Comment by gsakkis** [comment]
@wjones127 thanks! Btw I tried to guard it with a `threading.Lock()` shared by all threads but I'm getting the same error.

**Comment by benrapport** [comment]
Is there any timeline on when this will be implemented? Also what would you recommend for the interim to avoid duplication when adding a block of new embeddings?
